File: day6/main.hs
import Data.List ( nub, tails )


windows' :: Int -> [a] -> [[a]]
windows' n xs = map (take n) $ tails xs

windows :: Int -> [a] -> [[a]]
windows n xs = take (length xs - n + 1) (windows' n xs)

solve :: Int -> String -> Int
solve n xs =
  (+) n $
    length $
      takeWhile
        (== 1)
        (map (\ls -> if length (nub ls) == length ls then 0 else 1) $ windows n xs)

main :: IO ()
main = do
  contents <- readFile "./inp.txt"
  print $ solve 4 contents
  print $ solve 14 contents

File: day6/main.rs
fn has_dups(xs: &[u8]) -> bool {
    !(xs.iter().collect::<std::collections::HashSet<_>>().len() == xs.len())
}
fn main() {
    println!("{} {}", 
        std::fs::read("inp.txt").unwrap().windows(4).take_while(|x|has_dups(x)).count() + 4
    ,   std::fs::read("inp.txt").unwrap().windows(14).take_while(|x|has_dups(x)).count() + 14
    );
}